xml/value.go

```diff
@@ -0,0 +1,329 @@
+package xml
+
+import (
+	"encoding/base64"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+)
+
+// Value represents an XML Value type
+// XML Value types: Object, Array, Map, String, Number, Boolean.
+type Value struct {
+	w       writer
+	scratch *[]byte
+
+	// xml start element is the associated start element for the Value
+	startElement StartElement
+
+	// indicates if the Value represents a flattened shape
+	isFlattened bool
+}
+
+// newValue returns a new Value encoder. newValue does NOT write the start element tag
+func newValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	return Value{
+		w:            w,
+		scratch:      scratch,
+		startElement: startElement,
+	}
+}
+
+// newWrappedValue writes the start element xml tag and returns a Value
+func newWrappedValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	writeStartElement(w, startElement)
+	return Value{w: w, scratch: scratch, startElement: startElement}
+}
+
+// writeStartElement takes in a start element and writes it.
+// It handles namespace, attributes in start element.
+func writeStartElement(w writer, el StartElement) error {
+	if el.isZero() {
+		return fmt.Errorf("xml start element cannot be nil")
+	}
+
+	w.WriteRune(leftAngleBracket)
+
+	if len(el.Name.Space) != 0 {
+		w.WriteString(el.Name.Space)
+		w.WriteRune(colon)
+	}
+	w.WriteString(el.Name.Local)
+
+	for _, attr := range el.Attr {
+		w.WriteRune(' ')
+		buildAttribute(w, &attr)
+	}
+
+	w.WriteRune(rightAngleBracket)
+
+	return nil
+}
+
+// buildAttribute writes an attribute from a provided Attribute
+// For a namespace attribute, the attr.Name.Space must be defined as "xmlns".
+// https://www.w3.org/TR/REC-xml-names/#NT-DefaultAttName
+func buildAttribute(w writer, attr *Attr) {
```

Comment:
should this be named `writeAttribute`?  Its not actually building anything, just attributes the value to the writer.